src/Nethermind/Nethermind.JsonRpc/WebSockets/JsonRpcSocketsClient.cs

```diff
@@ -126,16 +129,14 @@ public virtual async Task<int> SendJsonRpcResult(JsonRpcResult result)
                         {
                             if (!isFirst)
                             {
-                                await _handler.SendRawAsync(_jsonComma, false);
-                                singleResponseSize += 1;
+                                await resultData.WriteAsync(_jsonComma);
                             }
-
                             isFirst = false;
-                            singleResponseSize += await SendJsonRpcResultEntry(entry, false);
+                            SendJsonRpcResultEntry(resultData, entry);
                             _ = _jsonRpcLocalStats.ReportCall(entry.Report);
 
                             // We reached the limit and don't want to responded to more request in the batch
-                            if (!_jsonRpcContext.IsAuthenticated && singleResponseSize > _maxBatchResponseBodySize)
+                            if (!_jsonRpcContext.IsAuthenticated && resultData.WrittenBytes > _maxBatchResponseBodySize)
```

Comment:
This feature is fundamentally broken. Consider the following scenarios where we are sending `JsonRpcResult`s and each value in `<>` represents the number of bytes:

```
Scenario 1: [ <100>, <100>, <999999> ] (Collection)
Scenario 2: { items: [ <999999>, <999999>, <999999>, ... ] } (Single)
```

On `1`, if we limit set `_maxBatchResponseBodySize` to `300`, that is, we limit the response size to 300 bytes, we will fully send the last element of the collection, sending a total of `100 + 100 + 9999999` bytes, which exceeds the limit.

On `2`, since we're dealing with a single element, we do not perform any checks, despite the element containing a collection of items that could exceed any possible limit.

Reply:
A proper solution would require the underlying stream (`resultData`) to stop as soon as the limit is reached, just by taking a look at the number of bytes written. A decorator like `LimitedStream(Stream stream, long writeAtMostBytes)` would suffice.

Reply:
There are tests (ignored for now) that show how this feature is currently broken